File: internal/domeneshop/invoices.go
package domeneshop

import (
	"context"
	"fmt"
	"net/url"
)

type Invoice struct {
	ID         int     `json:"id"`
	Type       string  `json:"type"`
	Amount     float64 `json:"amount"`
	Currency   string  `json:"currency"`
	DueDate    string  `json:"due_date"`
	IssuedDate string  `json:"issued_date"`
	PaidDate   string  `json:"paid_date"`
	Status     string  `json:"status"`
	URL        string  `json:"url"`
}

// InvoiceStatuses lists the accepted values for the ListInvoices status filter.
// "settled" applies only to credit notes.
var InvoiceStatuses = []string{"unpaid", "paid", "settled"}

// ListInvoices lists invoices from the past three years, optionally filtered by
// status.
//
// The status query parameter is missing from the published OpenAPI spec but is
// supported; Domeneshop's own JavaScript client relies on it. See
// https://github.com/domeneshop/domeneshop.js (src/lib/invoices.ts).
func (c *Client) ListInvoices(ctx context.Context, status string) ([]Invoice, error) {
	query := url.Values{}
	if status != "" {
		query.Set("status", status)
	}
	var invoices []Invoice
	if _, err := c.request(ctx, "GET", "/invoices", query, nil, &invoices); err != nil {
		return nil, err
	}
	return invoices, nil
}

// GetInvoice retrieves a single invoice.
func (c *Client) GetInvoice(ctx context.Context, invoiceID int) (*Invoice, error) {
	var invoice Invoice
	if _, err := c.request(ctx, "GET", fmt.Sprintf("/invoices/%d", invoiceID), nil, nil, &invoice); err != nil {
		return nil, err
	}
	return &invoice, nil
}

File: src/domeneshop/resources/invoices.ts
import { assertDefined } from "../assert.js";
import type { HttpTransport } from "../http.js";

/** "settled" applies only to credit notes. */
export const INVOICE_STATUSES = ["unpaid", "paid", "settled"] as const;

export type InvoiceStatus = (typeof INVOICE_STATUSES)[number];

export interface Invoice {
  id: number;
  type: string;
  amount: number;
  currency: string;
  due_date: string;
  issued_date: string;
  paid_date: string;
  status: string;
  url: string;
}

export class InvoicesResource {
  constructor(private readonly transport: HttpTransport) {}

  /**
   * Only returns invoices from the past three years. The status parameter is
   * undocumented but supported — see domeneshop.js (src/lib/invoices.ts).
   */
  async list(status?: string): Promise<Invoice[]> {
    const { data } = await this.transport.request<Invoice[]>("GET", "/invoices", {
      query: { status },
    });
    return data ?? [];
  }

  async get(invoiceId: number): Promise<Invoice> {
    const { data } = await this.transport.request<Invoice>("GET", `/invoices/${String(invoiceId)}`);
    return assertDefined(data, "domeneshop: empty response for invoice");
  }
}
